main/modes/pushy/pushy.c

```diff
@@ -0,0 +1,524 @@
+    pushy->rainbowHues[4]  = RAINBOW_HUE_STEP *  4;
+    pushy->rainbowHues[5]  = RAINBOW_HUE_STEP *  5;
+    pushy->rainbowHues[6]  = RAINBOW_HUE_STEP *  6;
+    pushy->rainbowHues[7]  = RAINBOW_HUE_STEP *  7;
+    pushy->rainbowHues[8]  = RAINBOW_HUE_STEP *  8;
+    pushy->rainbowHues[9]  = RAINBOW_HUE_STEP *  9;
+    pushy->rainbowHues[10] = RAINBOW_HUE_STEP * 10; // never actually used, as this is redirected to pushy->colors[10]
+    // clang-format on
+
+    shuffleColors();
+}
+
+/**
+ * This function is called when the mode is exited. It deinitializes variables and frees all memory.
+ */
+static void pushyExitMode(void)
+{
+    // Save score to NVS
+    if (pushy->lastSaveCounter != pushy->counter)
+    {
+        writeNvs32(pushyCounterKey, (int32_t)pushy->counter);
+    }
+
+    // Free the font
+    freeFont(&pushy->sevenSegment);
+
+    // Free everything else
+    free(pushy);
+}
+
+/**
+ * @brief This function is called periodically and frequently. It will either draw the menu or play the game, depending
+ * on which screen is currently being displayed
+ *
+ * @param elapsedUs The time that has elapsed since the last call to this function, in microseconds
+ */
+static void pushyMainLoop(int64_t elapsedUs)
+{
+    pushy->buttonPushedUs += elapsedUs;
+
+    readButton();
+
+    // If score has changed, save if last input was longer ago than our threshold or if score is a multiple of 100
+    if (pushy->lastSaveCounter != pushy->counter && pushy->buttonPushedUs > IDLE_SECONDS_UNTIL_SAVE * 1000 * 1000)
+    {
+        saveMemory();
+    }
+
+    // Draw "unlit" seven-segment displays
+    drawText(&pushy->sevenSegment, pushy->colors[NUM_PUSHY_COLORS - 1], pushy->eights,
+             (TFT_WIDTH - pushy->eightsWidth) / 2, (TFT_HEIGHT - pushy->sevenSegment.height) / 2);
+
+    // Draw "lit" segments
+    char counterStr[NUM_DIGITS + 1];
+    snprintf(counterStr, NUM_DIGITS + 1, "%*" PRIu32, NUM_DIGITS, pushy->counter);
+
+    updateEffects(counterStr);
+    updateFire();
+
+    displayCounter(counterStr);
+    displayFire();
+
+    setLeds(pushy->boxleds, CONFIG_NUM_LEDS);
+}
+
+/**
+ * This function is called when the display driver wishes to update a
+ * section of the display.
+ *
+ * @param disp The display to draw to
+ * @param x the x coordinate that should be updated
+ * @param y the x coordinate that should be updated
+ * @param w the width of the rectangle to be updated
+ * @param h the height of the rectangle to be updated
+ * @param up update number
+ * @param numUp update number denominator
+ */
+static void pushyBackgroundDrawCallback(int16_t x, int16_t y, int16_t w, int16_t h, int16_t up, int16_t upNum)
+{
+    // Use TURBO drawing mode to draw individual pixels fast
+    SETUP_FOR_TURBO();
+
+    // Draw a grid
+    for (int16_t yp = y; yp < y + h; yp++)
+    {
+        for (int16_t xp = x; xp < x + w; xp++)
+        {
+            TURBO_SET_PIXEL(xp, yp, c000);
+        }
+    }
```

Comment:
This seems like more of a solid color than a grid. Probably a better job for `fillDisplayArea()` but I don't think performance matters too much in this mode.